testnet/stacks-node/src/nakamoto_node/signer_coordinator.rs

```diff
@@ -0,0 +1,436 @@
+        }
+
+        let reward_cycle_id = burnchain
+            .block_height_to_reward_cycle(burn_tip.block_height)
+            .expect("FATAL: tried to initialize coordinator before first burn block height");
+
+        let block_proposal = BlockProposal {
+            block: block.clone(),
+            burn_height: burn_tip.block_height,
+            reward_cycle: reward_cycle_id,
+        };
+
+        let block_proposal_message = SignerMessageV0::BlockProposal(block_proposal);
+        debug!("Sending block proposal message to signers";
+            "signer_signature_hash" => %block.header.signer_signature_hash(),
+        );
+        Self::send_miners_message::<SignerMessageV0>(
+            &self.message_key,
+            sortdb,
+            burn_tip,
+            stackerdbs,
+            block_proposal_message,
+            MinerSlotID::BlockProposal,
+            self.is_mainnet,
+            &mut self.miners_session,
+            election_sortition,
+        )
+        .map_err(NakamotoNodeError::SigningCoordinatorFailure)?;
+        counters.bump_naka_proposed_blocks();
+
+        #[cfg(test)]
+        {
+            info!(
+                "SignerCoordinator: sent block proposal to .miners, waiting for test signing channel"
+            );
+            // In test mode, short-circuit waiting for the signers if the TEST_SIGNING
+            //  channel has been created. This allows integration tests for the stacks-node
+            //  independent of the stacks-signer.
+            if let Some(signatures) =
+                crate::tests::nakamoto_integrations::TestSigningChannel::get_signature()
+            {
+                debug!("Short-circuiting waiting for signers, using test signature");
+                return Ok(signatures);
+            }
+        }
+
+        self.get_block_status(
+            &block.header.signer_signature_hash(),
+            &block.block_id(),
+            chain_state,
+            sortdb,
+            burn_tip,
+            counters,
+        )
+    }
+
+    /// Get the block status for a given block hash.
+    /// If we have not yet received enough signatures for this block, this
+    /// method will block until we do. If this block shows up in the staging DB
+    /// before we have enough signatures, we will return the signatures from
+    /// there. If a new burnchain tip is detected, we will return an error.
+    fn get_block_status(
+        &self,
+        block_signer_sighash: &Sha512Trunc256Sum,
+        block_id: &StacksBlockId,
+        chain_state: &mut StacksChainState,
+        sortdb: &SortitionDB,
+        burn_tip: &BlockSnapshot,
+        counters: &Counters,
+    ) -> Result<Vec<MessageSignature>, NakamotoNodeError> {
+        let (lock, cvar) = &*self.blocks;
+        let mut blocks = lock.lock().expect("FATAL: failed to lock block status");
+
+        loop {
+            let (guard, timeout_result) = cvar
```

Comment:
This is another example of something we should help callers avoid.  Low-level synchronization like waiting on a condition variable can be hidden within a `Comms` struct.

Reply:
Fixed in `StackerDBListenerComms`.